### PR TITLE
Interaction Modal Support

### PR DESCRIPTION
**Pull request by VelvetThePanda**
# Summary
Adds support for Discord's new form input/modal interaction flow. 

# Details
This isn't even in a beta yet, so I'm unable to test. 

I've added the requisites of a new interaction type, event args, dispatch, etc.

# Changes proposed
```diff
+ Add new interaction resp type: Modal (9)
+ Add new interaction type: ModalSubmit (5)
+ Add Title and Description to DiscordInteractionResposneBuilder
+ Add Title and Description to REST payloads
+ Add new event args, with convience dictionary for pulling repsonse data
+ Add ModalSubmitted event to DiscordClient & DiscordShardedClient
```

# Notes
**This isn't even in beta, so testing will be postponed.**

Also I added a single itneractivity method, because you're probably better off just hooking an event handler, but meh.

### CODE REVIEW

**Comment by VelvetThePanda** [comment]
Docs have been merged. Review would be nice.

File: DSharpPlus.Test/TestBot.cs [hunk 1]
```diff
@@ -94,6 +94,8 @@ public TestBot(TestBotConfig cfg, int shardid)
             this.Discord.ChannelDeleted += this.Discord_ChannelDeleted;
 
             this.Discord.InteractionCreated += this.Discord_InteractionCreated;
+            this.Discord.ComponentInteractionCreated+= this.Discord_ModalCheck;
```
Comment (tenhobi) [inline DSharpPlus.Test/TestBot.cs:97]: Just a nit.

```suggestion
            this.Discord.ComponentInteractionCreated += this.Discord_ModalCheck;
```

File: DSharpPlus.Interactivity/InteractivityExtension.cs [hunk 2]
```diff
@@ -114,6 +114,14 @@ public async Task<ReadOnlyCollection<PollEmoji>> DoPollAsync(DiscordMessage m, I
             return new ReadOnlyCollection<PollEmoji>(res.ToList());
         }
 
+        /// <summary>
+        /// Waits for a user to submit a modal.
+        /// </summary>
+        /// <param name="modal_id">The id of the modal to wait for.</param>
+        /// <param name="timeoutOverride">Override the timeout period in <see cref="InteractivityConfiguration"/>.</param>
+        public Task<InteractivityResult<ModalSubmitEventArgs>> WaitForModalAsync(string modal_id, TimeSpan? timeoutOverride = null)
```
Comment (OoLunar) [inline DSharpPlus.Interactivity/InteractivityExtension.cs:122]: Nitpick: camelCase argument name

File: DSharpPlus/Entities/Interaction/Components/TextInputComponent.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        /// </summary>
+        /// <param name="label">The label for the field, placed above the input itself.</param>
+        /// <param name="customId">The ID of this field.</param>
+        /// <param name="placeholder">Placeholder text for the field.</param>
+        /// <param name="value">A pre-filled value for this field.</param>
+        /// <param name="required">Whether this field is required.</param>
+        /// <param name="style">The style of this field. A single-ling short, or multi-line paragraph.</param>
+        /// <param name="min_length">The minimum input length.</param>
+        /// <param name="max_length">The maximum input length. Must be greater than the minimum, if set.</param>
+        public TextInputComponent(string label, string customId, string placeholder = null, string value = null, bool required = true, TextInputStyle style = TextInputStyle.Short, int min_length = 0, int? max_length = null)
```
Comment (OoLunar) [inline DSharpPlus/Entities/Interaction/Components/TextInputComponent.cs:90]: Nitpick: camelCase argument name

File: DSharpPlus/Entities/Interaction/DiscordInteractionResponseBuilder.cs [hunk 4]
```diff
@@ -136,6 +146,34 @@ public DiscordInteractionResponseBuilder AddComponents(IEnumerable<DiscordAction
...
+
+            this.Title = title;
+            return this;
+        }
+
+        /// <summary>
+        /// If responding with a modal, sets the custom id for the modal.
+        /// </summary>
+        /// <param name="title"></param>
+        /// <returns></returns>
```
Comment (OoLunar) [inline DSharpPlus/Entities/Interaction/DiscordInteractionResponseBuilder.cs:167]: Nitpick: Missing doc

**Review by OoLunar — APPROVED** [review]

**Comment by VelvetThePanda** [comment]
The argument names are just a consistency thing with the rest of the codebase

**Comment by VelvetThePanda** [comment]
@kuylar if you could verify that value works, I'll merge after

**Comment by kuylar** [comment]
Just got around to testing it, works fine without any problems

**Comment by VelvetThePanda** [comment]
If this breaks, @Emzi0767 just know it was @kuylar

**Comment by Emzi0767** [comment]
🥚

**Comment by Naamloos** [comment]
🥚

**Pull request merged**